packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+        std::vector<int> bndry_recvcounts(comm->getSize(),0);
+        Teuchos::ArrayView<int> bndry_recvcounts_view = Teuchos::arrayViewFromVector(bndry_recvcounts);
+        Zoltan2::AlltoAllv<int> (*comm, *env, is_bndry_send_view, recvcounts_view, is_bndry_recv, bndry_recvcounts_view);
+
+	//add together the flags to compute how many boundary vertices are in the second ghost layer
+        int boundaryverts = 0;
+        for(int i = 0; i < is_bndry_recv.size(); i++){
+          boundaryverts += is_bndry_recv[i];
+        }
+	//this cout is guarded by a verbose check.
+        std::cout<<comm->getRank()<<": "<<boundaryverts<<" boundary verts out of "<<n2Ghosts<<" verts in 2GL\n";
+      }
+
+      //create ArrayViews from the local vectors, to prevent changes happening in the coloring
+      //
+      //local CSR representation for the owned vertices: 
+      Teuchos::ArrayView<const lno_t> local_adjs_view = Teuchos::arrayViewFromVector(local_adjs);
+      //NOTE: the offset ArrayView was constructed earlier, and is up-to-date.
+      //
+      //first ghost layer CSR representation:
+      Teuchos::ArrayView<const offset_t> ghost_offsets = Teuchos::arrayViewFromVector(first_layer_ghost_offsets);
+      Teuchos::ArrayView<const lno_t> ghost_adjacencies = Teuchos::arrayViewFromVector(local_ghost_adjs);
+
+      //these ArrayViews contain LIDs that are ghosted remotely,
+      //and the Process IDs that have those ghost copies.
+      //An LID may appear in the exportLIDs more than once.
+      Teuchos::ArrayView<const lno_t> exportLIDs = importer->getExportLIDs();
+      Teuchos::ArrayView<const int> exportPIDs = importer->getExportPIDs();
+      
+      //construct a quick lookup datastructure to map from LID to a
+      //list of PIDs we need to send data to.
+      std::unordered_map<lno_t, std::vector<int>> procs_to_send;
+      for(int i = 0; i < exportLIDs.size(); i++){
+        procs_to_send[exportLIDs[i]].push_back(exportPIDs[i]);
+      }
+
+      //call the coloring algorithm
+      twoGhostLayer(nVtx, nVtx+nGhosts, local_adjs_view, offsets, ghost_adjacencies, ghost_offsets,
+                    femv, ownedPlusGhosts, rand, owners2, mapWithCopies, procs_to_send);
+      
+      //copy colors to the output array
+      ArrayRCP<int> colors = solution->getColorsRCP();
+      auto femvdata = femv->getData(0);
+      for(size_t i=0; i<nVtx; i++){
+         colors[i] = femvdata[i];
+      }
+
+    }
+//  private:
+
+    //This is the coloring logic for all 2GL-based methods.
+    //
+    //INPUT ARGS:
+    //  
+    //  n_local: the number of local owned vertices
+    //
+    //  n_total: the number of local owned vertices plus first layer ghosts
+    //
+    //  adjs: the CSR adjacencies for the graph, includes first and second
+    //        layer ghosts
+    //
+    //  offsets: the CSR offsets for the graph, has owned offsets into adjacencies
+    //
+    //  ghost_adjs: the adjacency information for only the second ghost layer
```

Comment:
Better comment:  adjacency information for first-layer ghost vertices; includes all neighbors (owned, first-layer ghost, second-layer ghost) for each first layer ghost ??

Reply:
That is more accurate. I'll update the comment.